README.md
## About

A node.js module allows you easily maintain simple data structure on Amazon S3 service (Use [aws2js](https://github.com/SaltwaterC/aws2js) for S3 access):

1. KeyList

Maintains a list of string keys. You can:

* Add a key / keys into the list
* Fetch a list of keys on list
* Delete a key / keys / all keys from list
* Tell whether a key is / keys are on list

2. ObjectDB

Maintains a key value store. You can:

* Store a value / object under a string key
* Get the value / object of a string key
* Delete a key
* Tell whether a key exists

## Installation

Either manually clone this repository into your node_modules directory, or the recommended method:

> npm install aws2js

## Getting Started

### Initailization

```javascript
var s3db=require('s3db'); //--- load s3db module
var db=new s3db('your aws_access key id','your aws secrect access key');
```

### KeyList

```javascript
var kl=db.open('KeyList','s3://mybucket.com/path_to_my_key_list/my_key_list'); //--Open a keylist

//--- fetch keys on list
kl.get(function(keys){ //--- callback will be called when operation completed without error.
	console.log(keys); //-- first 100 keys
	if(kl.is_truncated==true){ //--- if not all keys return in one request (default fetch only first 1000 keys), then continue to fetch the rest keys on the list.
		kl.get(function(keys){
			console.log(keys); //-- next 1000 keys
		},function(error){
			console.log(error);
		});
	}
},function(error){ //--- callback will be called when error occurs
	console.log(error);
});

//--- delete all keys on list

kl.del_all(function(){  //--- callback will be called when operation completed without error.
	console.log('Operation completed'); //-- next 1000 keys
},function(error){
	console.log(error);
});

```

### ObjectDB

```javascript
var odb=db.open('ObjectDB','s3://mybucket.com/path_to_obejct_db/my_object_db'); //--Open a ObjectDB

//--- store an object / hash under key 'mykey'
var obj={a:1,b:2};

odb.put('mykey',obj,function(){  //--- callback will be called when operation completed without error. 
	console.log('Operation completed');
	
	//-- get value under key 'mykey'
	odb.get('mykey',function(obj){  //--- callback will be called when operation completed without error. 
		console.log(obj); //--- will print out: {a:1,b:2}
	},function(e){	//--- callback will be called when error occurs
		console.log(e);
	});
	
},function(e){	//--- callback will be called when error occurs
	console.log(e);
});

```

## `Promise` Interface Support

Along with callback style API, `Promise` stlyle asynchronous programming pattern is also support (Based on JavaScript Library [Q](https://github.com/kriskowal/q/)).


```javascript

kl.get() //---get full key list
.then(kl.del) //-- del all key returned by kl.get()
.then(kl.get) //-- fetch key list again
.then(function(r){
	console.log(obj); //--- will print out: []
})fail(function(error){
	console.log(error); //--- print out error during the whole process
});

```

```javascript
//---Load underscore library for function bind
var _=require('underscore');

//--- store an object / hash under key 'mykey'
var obj={a:1,b:2};

odb.put('mykey',obj).then(_.partial(odb.get,
								 'mykey' //--- paramter for get method
)).then(function(obj){
	console.log(obj); //--- will print out: {a:1,b:2}
}).fail(function(error){
	console.log(error); //--- print out error during the whole process
});

```

## API Reference

### s3db

##### Method init(aws_access_key_id,aws_secrect_access_key,endpoint_prefix)

endpoint_prefix: default to ''. Example: 's3-us-west-1' The endpoint list is available [here](http://docs.amazonwebservices.com/general/latest/gr/index.html?rande.html#s3_region).

##### Method open(module,uri)
module: availble values: 'KeyList', 'ObjectDB'

### KeyList

##### Method get(ok_func,error_func,limit,marker)
ok_func: callback. Will be called if operation completed withour error. Optional if use `Promise` API.
ok_func: callback. Will be called if error occurs. Optional if use `Promise` API.
limit: Max. number of keys to fetch. Optional, default to 1000. A value larger than 1000 will be ignored.
marker: Fetch the keys from the marker. For pagination purpose. Optional.

##### Method reset()
Reset internal marker so that get method will return result form the first key rather then where the result is truncated during last method get call.

##### Method put(key,ok_func,error_func)
key: Can be a string key or an array of keys that should be put on list

##### Method del(key,ok_func,error_func)
key: Can be a string key or an array of keys that should be deleted from list

##### Method del_all(ok_func,error_func)

##### Method on_list(key,yes_func,no_func,error_func)
key: Can be a string key or an array of keys
yes_func: callback will be called if all keys are on list


### ObjectDB

##### Method get(key,ok_func,error_func)
key: String. key of the value / object

##### Method put(key,obj,ok_func,error_func)
key: String. key of the value / object
obj: Object. The object should be store under the key.

##### Method del(key,ok_func,error_func)
key: String. key of the value / object

##### Method is_set(key,yes_func,no_func,error_func)
key: String. key of the value / object
yes_func: callback will be called if the key exists




CHANGES.md
## 0.0.3

 - fix: KeyList does not set next_marker properly when delimiter is set to '' or null
 
## 0.0.2

 - Now all methods (Object.get,KeyList.get etc...) are always bound to orginal object
 - Now Object.get will return null if key not exsits
 - Upgrade aws2js version to 0.8.3
 - Upgrade Q versio to 0.9.6

test/ObjectDB.js
var assert = require('assert');
var s3db=require('../');
s3db.setCredentials(process.env.AWS_ACCEESS_KEY_ID, process.env.AWS_SECRET_ACCESS_KEY);

var objdb=new s3db.load('ObjectDB',process.env.S3_DB_URI);

lib/s3db.js
var _=require('underscore'),
aws2js=require('aws2js'),
URLUtils=require('url'),
querystring=require('querystring'),
crypto=require('crypto');

var ObjectDB=require('./ObjectDB'),
KeyList=require('./KeyList');

function init(aws_access_key_id,aws_secrect_access_key,endpoint_prefix)
{
	if(!_.isString(aws_access_key_id)) aws_access_key_id='';
	if(!_.isString(aws_secrect_access_key)) aws_secrect_access_key='';
	aws_access_key_id=aws_access_key_id.trim();
	aws_secrect_access_key=aws_secrect_access_key.trim();
	if(aws_access_key_id=='' || aws_secrect_access_key=='') throw new Error('s3db init: missing aws_access_key_id or aws_secrect_access_key parameters.');
	this.aws_access_key_id=aws_access_key_id;
	this.aws_secrect_access_key=aws_secrect_access_key;
	this.s3=aws2js.load('s3',this.aws_access_key_id,this.aws_secrect_access_key);
	if(endpoint_prefix && _.isString(endpoint_prefix)) this.endpoint_prefix=endpoint_prefix;
	else this.endpoint_prefix=null;
}

module.exports=init;

init.prototype.open=function(module,uri){
	switch(module)
	{
		case 'ObjectDB' : return new ObjectDB(this.s3,uri,this.endpoint_prefix);
		case 'KeyList' : return new KeyList(this.s3,uri,this.endpoint_prefix);
	}
	throw new Error('Unknow s3db module:'+module);
};

lib/ObjectDB.js
var _=require('underscore'),
URLUtils=require('./URLUtils'),
s3db=require('./s3db'),
Q=require('q');

function init(s3,uri,endpoint_prefix)
{
	if(!s3 || !uri) throw new Error('init@ObjectDB@s3db: s3 and uri parameters cannot be empty');
	this.s3=s3;
	var info=URLUtils.parse_s3_uri(uri);
	if(!info) throw new Error('init@ObjectDB@s3db: Invalid S3 URI: '+uri);
	if(!endpoint_prefix) s3.setBucket(info['bucket']);
	else s3.setBucket(info['bucket']+'.'+endpoint_prefix);
	this.path=URLUtils.add_ending_slash(info['path']);
	
	this.bind_methods();
}
module.exports=init;

init.prototype.bind_methods=function(){
	this.get=_.bind(this.get,this);
	this.put=_.bind(this.put,this);
	this.del=_.bind(this.del,this);
	this.is_set=_.bind(this.is_set,this);
};

init.prototype.get=function(key,ok_func,error_func){
	var deferred = Q.defer();
	try{
		if(!key || !_.isString(key)) throw new Error('Invalid or empty Key');
		this.s3.get(this.path+key,'buffer',function(error,response){
			try{
				if(error) {
					if(error['code'] && error['code']=='404') {
						if(ok_func && _.isFunction(ok_func)) ok_func(null,response['headers']);
						deferred.resolve(null);
						return;
					}
					throw error;
				}
				if(!response.buffer || !Buffer.isBuffer(response.buffer)) throw new Error('Empty or invalid Buffer Object Returned');
				var data=response.buffer.toString('utf-8');
				data=JSON.parse(data);
				if(ok_func && _.isFunction(ok_func)) ok_func(data,response['headers']);
				deferred.resolve(data);
			}catch(e){
				e.key=key;
				if(error_func && _.isFunction(error_func)) error_func(e);
				deferred.reject(e);
			}
		});	
	}catch(e){
		e.key=key;
		if(error_func && _.isFunction(error_func)) error_func(e);
		deferred.reject(e);
	}
	return deferred.promise;
};

init.prototype.put=function(key,obj,ok_func,error_func){
	var deferred = Q.defer();
	try{
		if(!key || !_.isString(key) || !obj) throw new Error('Invalid or empty Key Or empty data');
		obj=new Buffer(JSON.stringify(obj),'utf-8');
		this.s3.putBuffer(this.path+key,obj,'private',{
			'Content-Type':'application/json'
		},function(error,response){
			try{
				if(error) throw error;
				if(ok_func && _.isFunction(ok_func)) ok_func(true);
				deferred.resolve(true);
			}catch(e){
				e.key=key;
				if(error_func && _.isFunction(error_func)) error_func(e);
				deferred.reject(e);
			}
		});	
	}catch(e){
		e.key=key;
		if(error_func && _.isFunction(error_func)) error_func(e);
		deferred.reject(e);
	}
	return deferred.promise;
};

init.prototype.del=function(key,ok_func,error_func){
	var deferred = Q.defer();
	try{
		if(!key || !_.isString(key)) throw new Error('Invalid or empty Key');
		this.s3.del(this.path+key,function(error,response){
			try{
				if(error) throw error;
				if(ok_func && _.isFunction(ok_func)) ok_func(true);
				deferred.resolve(true);
			}catch(e){
				e.key=key;
				if(error_func && _.isFunction(error_func)) error_func(e);
				deferred.reject(e);
			}
		});	
	}catch(e){
		e.key=key;
		if(error_func && _.isFunction(error_func)) error_func(e);
		deferred.reject(e);
	}
	return deferred.promise;
};

init.prototype.is_set=function(key,yes_func,no_func,error_func){
	var deferred = Q.defer();
	try{
		if(!key || !_.isString(key)) throw new Error('Empty & invalid Key');
						
		this.s3.head(this.path+key,_.bind(function(error,response){
			if(error) {
				if(error['code'] && error['code']=='404') {
					if(no_func && _.isFunction(no_func)) no_func();
					deferred.resolve(false);
					return;	
				}
				error.key=key;
				if(error_func && _.isFunction(error_func)) error_func(error);
				deferred.reject(error);
				return;
			}
			if(yes_func && _.isFunction(yes_func)) yes_func();
			deferred.resolve(true);	
		},this));		
	}catch(e){
		e.key=key;
		if(error_func && _.isFunction(error_func)) error_func(e);
		deferred.reject(e);
	}
	return deferred.promise;
};



lib/KeyList.js
var _=require('underscore'),
URLUtils=require('./URLUtils'),
s3db=require('./s3db'),
Q=require('q'),
crypto=require('crypto');

exports=module.exports=init;

function init(s3,uri,endpoint_prefix)
{
	if(!s3 || !uri) throw new Error('init@ObjectDB@s3db: s3 and uri parameters cannot be empty');
	this.s3=s3;
	var info=URLUtils.parse_s3_uri(uri);
	if(!info) throw new Error('init@ObjectDB@s3db: Invalid S3 URI: '+uri);
	if(!endpoint_prefix) s3.setBucket(info['bucket']);
	else s3.setBucket(info['bucket']+'.'+endpoint_prefix);
	this.prefix=URLUtils.remove_leading_slash(URLUtils.add_ending_slash(info['path']));
	this.path=URLUtils.add_ending_slash(info['path']);
	
	this.next_marker='';
	this.is_truncated=false;
	this.limit=0;
	this.delimiter='';
	
	this.bind_methods();
}

init.prototype.bind_methods=function(){
	this.reset=_.bind(this.reset,this);
	this.get=_.bind(this.get,this);
	this.put=_.bind(this.put,this);
	this.del=_.bind(this.del,this);
	this.del_all=_.bind(this.del_all,this);
	this.on_list=_.bind(this.on_list,this);
};

init.prototype.default_limit=1000;
init.prototype.default_delimiter='/';

init.prototype.reset=function(){
	var deferred = Q.defer();
	this.next_marker='';
	this.is_truncated=false;
	this.limit=0;
	this.delimiter='';
	deferred.resolve(true);
	return deferred.promise;
};

init.prototype.get=function(ok_func,error_func,limit,marker){
	var deferred = Q.defer();
	try{

		var params={};
	
		var delimiter=(this.is_truncated && this.delimiter)?this.delimiter:this.default_delimiter;
		if(delimiter) params['delimiter']=delimiter;
		
		if(this.prefix) params['prefix']=this.prefix;
		
		if(!limit) limit=(this.is_truncated && this.limit) ? this.limit : this.default_limit;
		
		params['max-keys']=limit;
		
		if(!marker && this.is_truncated && this.next_marker) marker=this.next_marker;
		if(marker) params['marker']=marker;
		
		if(_.isEmpty(params)) params=null;
		
		this.s3.get('/',params,'xml',_.bind(function(error,response){
			try{
				if(error) throw error;
				this.is_truncated=response['IsTruncated']=='true'?true:false;
				this.next_marker=(response['NextMarker'] && typeof(response['NextMarker'])=='string')?response['NextMarker']:'';
				this.limit=(response['MaxKeys'] && typeof(response['MaxKeys'])=='string')?response['MaxKeys']:0;
				this.delimiter=(response['Delimiter'] && typeof(response['Delimiter'])=='string')?response['Delimiter']:'';
				var prefix=(response['Prefix'] && typeof(response['Prefix'])=='string')?response['Prefix']:'';
				
				var data=[];
				var key;
				
				if(response['Contents'] && response['Contents']['length'])
				{
					for(var i=0;i<response['Contents']['length'];i++)
					{
						if(!response['Contents'][i] || !response['Contents'][i]['Key'] || typeof(response['Contents'][i]['Key'])!='string') continue;
						key=response['Contents'][i]['Key'];
						if(!prefix) data.push(key);
						else data.push(key.substr(prefix.length));
					}
				}else if(response['Contents'] && response['Contents']['Key'] && typeof(response['Contents']['Key'])=='string'){
					key=response['Contents']['Key'];
					if(!prefix) data.push(key);
					else data.push(key.substr(prefix.length));
				}
				if(this.is_truncated && this.next_marker=='' && data['length']) this.next_marker=data[data.length-1];
				if(ok_func && _.isFunction(ok_func)) ok_func(data,this.is_truncated,this.next_marker,this.limit,this.delimiter);
				deferred.resolve(data);
			}catch(e){
				if(error_func && _.isFunction(error_func)) error_func(e);
				deferred.reject(e);
			}
		},this));	
	}catch(e){
		if(error_func && _.isFunction(error_func)) error_func(e);
		deferred.reject(e);
	}
	return deferred.promise;
};

init.prototype.put=function(key,ok_func,error_func){
	var main_deferred = Q.defer();
	try{
		if(!key) {
			main_deferred.resolve(true);
			return main_deferred.promise;
		}
		var keys;
		if(_.isArray(key)) {
			if(!key.length) {
				main_deferred.resolve(true);
				return main_deferred.promise;
			}
			keys=key;
		}else if(_.isString(key)) {
			keys=[key];
		}else throw new Error('Unknow `key` parameter type');

		var data=new Buffer('1','utf-8');
		
		var promises=_.map(keys,function(k){
			var deferred = Q.defer();

			this.s3.putBuffer(this.path+k,data,'private',{
				'Content-Type':'application/json'
			},function(error,response){
				if(error) deferred.reject(error);
				else deferred.resolve();	
			});
			return deferred.promise;
		},this);
		
		Q.all(promises).then(function(){
			if(ok_func && _.isFunction(ok_func)) ok_func();
			main_deferred.resolve(true);
		},function(e){
			if(error_func && _.isFunction(error_func)) error_func(e);
			main_deferred.reject(e);
		});
	
	}catch(e){
		if(error_func && _.isFunction(error_func)) error_func(e);
		main_deferred.reject(e);
	}
	return main_deferred.promise;
};

init.prototype.del=function(key,ok_func,error_func){
	var main_deferred = Q.defer();
	try{
		if(!key) {
			main_deferred.resolve(true);
			return main_deferred.promise;
		}
		var keys;
		if(_.isArray(key)) {
			if(!key.length) {
				main_deferred.resolve(true);
				return main_deferred.promise;
			}
			keys=key;
		}else if(_.isString(key)) {
			keys=[key];
		}else throw new Error('Unknow `key` parameter type');
		
		var tasks=[];
		if(keys.length<=this.default_limit) tasks.push(keys);
		else {
			for(var i=0;i<keys.length-this.default_limit;i+=this.default_limit)
				tasks.push(keys.slice(i,i+this.default_limit));
		}
		
		var promises=_.map(tasks,function(k_arr){
			var deferred = Q.defer();
			
			var req='<?xml version="1.0" encoding="UTF-8"?><Delete><Quiet>true</Quiet>';
			
			for(var i=0;i<k_arr.length;i++)
				req+='<Object><Key>'+_.escape(this.prefix+k_arr[i])+'</Key></Object>';
			
			req+='</Delete>';
			
			req=new Buffer(req,'utf-8');
			
			var hash=crypto.createHash('md5');
			hash.update(req);
			var md5=hash.digest('base64');
			
			this.s3.post('/?delete',{
				'Content-MD5':md5,
				'Content-Length':req.length
			},req,function(error,response){
				if(error) deferred.reject(error);
				else deferred.resolve();	
			});
			
			return deferred.promise;
		},this);
		
		Q.all(promises).then(function(){
			if(ok_func && _.isFunction(ok_func)) ok_func();
			main_deferred.resolve(true);
		},function(e){
			if(error_func && _.isFunction(error_func)) error_func(e);
			main_deferred.reject(e);
		});
	
	}catch(e){
		if(error_func && _.isFunction(error_func)) error_func(e);
		main_deferred.reject(e);
	}
	return main_deferred.promise;
};

init.prototype.del_all=function(ok_func,error_func){
	var deferred = Q.defer();
	try{
		var get_list_func=_.bind(function(){
			this.get().then(_.bind(this.del,this)).then(_.bind(function(){
				if(!this.is_truncated) {
					if(ok_func && _.isFunction(ok_func)) ok_func();
					deferred.resolve(true);
				}else get_list_func();
			},this)).fail(function(e){
				if(error_func && _.isFunction(error_func)) error_func(e);
				deferred.reject(e);
			});
		},this);
		get_list_func();	
	}catch(e){
		if(error_func && _.isFunction(error_func)) error_func(e);
		deferred.reject(e);
	}
	return deferred.promise;
};

init.prototype.on_list=function(key,yes_func,no_func,error_func){
	var main_deferred = Q.defer();
	try{
		if(!key) throw new Error('Empty Key');
		var keys;
		if(_.isArray(key)) {
			if(!key.length) throw new Error('Empty Key Array');
			keys=key;
		}else if(_.isString(key)) {
			keys=[key];
		}else throw new Error('Unknow `key` parameter type');
		
		key=null;
		
		var promises=_.map(keys,function(k){
			var deferred = Q.defer();
						
			this.s3.head(this.path+k,function(error,response){
				if(error) {
					error.key=k;
					deferred.reject(error);
				}else deferred.resolve();	
			});
			
			return deferred.promise;
		},this);
		
		Q.all(promises).then(function(){
			if(yes_func && _.isFunction(yes_func)) yes_func();
			main_deferred.resolve(true);
		},function(e){
			if(e['code'] && e['code']=='404') {
				if(no_func && _.isFunction(no_func)) no_func(e.key,e);
				main_deferred.resolve(false);
				return;
			}
			if(error_func && _.isFunction(error_func)) error_func(e);
			main_deferred.reject(e);
		});
		
	}catch(e){
		if(error_func && _.isFunction(error_func)) error_func(e);
		main_deferred.reject(e);
	}
	return main_deferred.promise;
};

